examples/dnn_introduction3_ex.cpp

```diff
@@ -0,0 +1,180 @@
+/*
+    This is an example illustrating the use of the deep learning tools from the
+    dlib C++ Library.  I'm assuming you have already read the dnn_introduction_ex.cpp and
+    the dnn_introduction2_ex.cpp examples.  So in this example program I'm going to go
+    over a transfer learning example, which includes:
+        - Defining a layer visitor to modify the learning rate for fine-tuning
+        - Using pretrained layers of a network for another task
+*/
+
+#include <dlib/dnn.h>
+#include <iostream>
+
+// This header file includes a generic definition of the most common ResNet architectures
+#include "resnet.h"
+
+using namespace std;
+using namespace dlib;
+
+// In this simple example we will show how to load a pretrained network
+// and use it for a different task.  In particular, we will load a ResNet50
+// trained on ImageNet and use it as a pretrained backbone for some metric
+// learning task
+
+// Let's start by defining a network that will use the ResNet50 backbone from
+// resnet.h
+namespace model
+{
+    template<template<typename> class BN>
+    using net_type = loss_metric<
+        fc_no_bias<128,
+        avg_pool_everything<
+        typename resnet<BN>::template backbone_50<
+        input_rgb_image
+        >>>>;
+
+    using train = net_type<bn_con>;
+    using infer = net_type<affine>;
+}
+
+// Next, we define a layer visitor that will modify the learning rate of a network.
+// In particular, we have to define it for layers that implement the following methods:
+// - set_learning_rate_multiplier
+// - set_bias_learning_rate_multiplier
+class visitor_learning_rate_multiplier
+{
+public:
+
+    visitor_learning_rate_multiplier(double new_learning_rate_multiplier_) :
+        new_learning_rate_multiplier(new_learning_rate_multiplier_) {}
+
+    template <typename T>
+    void set_learning_rate_multiplier(T&) const
+    {
+        // ignore other layer detail types
+    }
+
+    // convolutional layers
+    template <long nf, long nr, long nc, int sx, int sy>
+    void set_learning_rate_multiplier(con_<nf,nr,nc,sx,sy>& l) const
+    {
+        l.set_learning_rate_multiplier(new_learning_rate_multiplier);
+        l.set_bias_learning_rate_multiplier(new_learning_rate_multiplier);
+    }
+
+    // batch normalization layers
+    template <layer_mode mode>
+    void set_learning_rate_multiplier(bn_<mode>& l) const
+    {
+        l.set_learning_rate_multiplier(new_learning_rate_multiplier);
+        l.set_bias_learning_rate_multiplier(new_learning_rate_multiplier);
+    }
+
+    // fully connected layers (with and without bias)
+    template<unsigned long num_outputs, fc_bias_mode bias_mode>
+    void set_learning_rate_multiplier(fc_<num_outputs, bias_mode>& l) const
+    {
+        l.set_learning_rate_multiplier(new_learning_rate_multiplier);
+        if (bias_mode == FC_HAS_BIAS)
+            l.set_bias_learning_rate_multiplier(new_learning_rate_multiplier);
+    }
```

Comment:
Instead of explicitly listing the layers that have learning rate multipliers, you could do what `dlib::get_learning_rate_multiplier()` does.  In fact, defining a `dlib::set_learning_rate_multiplier()` and just calling that would be good.  Same for the bias. I just looked and I'm a little surprised I left these set methods out :|

Reply:
Oh, I see. I've just look at that code and it seems a bit over my C++ skills, so it might take some time to figure it out :)

Reply:
I just mean this stuff:
```
    namespace impl 
    {    
        template <typename T, typename int_<decltype(&T::get_learning_rate_multiplier)>::type = 0> 
        double get_learning_rate_multiplier (
            const T& obj, 
            special_
        ) { return obj.get_learning_rate_multiplier(); }

        template <typename T>
        double get_learning_rate_multiplier ( const T& , general_) { return 1; } 
    }    
    template <typename T>
    double get_learning_rate_multiplier(const T& obj) { return impl::get_learning_rate_multiplier(obj, special_()); }
```
It uses SFINAE (https://en.wikipedia.org/wiki/Substitution_failure_is_not_an_error) to select which overload of `get_learning_rate_multiplier` to use. The key bit here is `int_<decltype(&T::get_learning_rate_multiplier)>::type`, which is an error if `T::get_learning_rate_multiplier` doesn't exist.

Reply:
Thank you! I am not very familiar with SFINAE, so I just modified it to something it seemed sensible, but doesn't work yet, I will learn how to do it :) This is what I tried, if you want to shame me :package: 

``` c++
  namespace impl
  {
      template <typename T, typename int_<decltype(&T::set_learning_rate_multiplier)>::type = 0>
      void set_learning_rate_multiplier (
          T& obj,
          special_,
          double new_learning_rate_multiplier
      ) { obj.set_learning_rate_multiplier(new_learning_rate_multiplier); }
  
      template <typename T>
      void set_learning_rate_multiplier (T& , general_) { }
  }
  template <typename T>
  void set_learning_rate_multiplier(
      T& obj,
      double new_learning_rate_multiplier
  )
  {
      impl::set_learning_rate_multiplier(obj, special_(), new_learning_rate_multiplier);
  }
```

Reply:
I found the mistake, I missed the `double` parameter in the general case... I will add now the missing functions :)